osquery/tables/networking/windows/routes.cpp

```diff
@@ -0,0 +1,168 @@
+*  All rights reserved.
+*
+*  This source code is licensed under the BSD-style license found in the
+*  LICENSE file in the root directory of this source tree. An additional grant
+*  of patent rights can be found in the PATENTS file in the same directory.
+*
+*/
+
+#define WIN32_LEAN_AND_MEAN
+
+#include <boost/algorithm/string/join.hpp>
+#include <osquery/tables.h>
+#include <string>
+#include <windows.h>
+#include <winsock2.h>
+
+#include <wS2tcpip.h>
+#include <ws2ipdef.h>
+
+#include <iphlpapi.h>
+#include <mstcpip.h>
+
+#include "osquery/core/conversions.h"
+#include "osquery/core/windows/wmi.h"
+
+#pragma comment(lib, "Iphlpapi.lib")
+#pragma comment(lib, "Ws2_32.lib")
+
+namespace osquery {
+namespace tables {
+
+PMIB_IPINTERFACE_TABLE getInterfaces(int type = AF_UNSPEC) {
+  DWORD dwRetVal = 0;
+  PMIB_IPINTERFACE_TABLE interfaceTable = nullptr;
+
+  dwRetVal = GetIpInterfaceTable(type, &interfaceTable);
+  if (dwRetVal != NO_ERROR) {
+    return nullptr;
+  }
+
+  return interfaceTable;
+}
+
+std::map<unsigned long, PIP_ADAPTER_INFO> getAdapterAddressMapping() {
+  std::map<unsigned long, PIP_ADAPTER_INFO> returnMapping;
+  DWORD dwBufLen = 0;
+  DWORD dwStatus = GetAdaptersInfo(NULL, &dwBufLen);
+
+  if (dwStatus == ERROR_BUFFER_OVERFLOW) {
+    PIP_ADAPTER_INFO pAdapterInfo = (PIP_ADAPTER_INFO)malloc(dwBufLen);
+    dwStatus = GetAdaptersInfo(pAdapterInfo, &dwBufLen);
+    if (dwStatus == S_OK) {
+      while (pAdapterInfo) {
+        returnMapping.insert(std::pair<unsigned long, PIP_ADAPTER_INFO>(
+            pAdapterInfo->Index, pAdapterInfo));
+        pAdapterInfo = pAdapterInfo->Next;
+      }
+    }
+  }
+
+  return returnMapping;
+}
+
+std::map<unsigned long, MIB_IPINTERFACE_ROW> getInterfaceRowMapping(
+    int type = AF_UNSPEC) {
+  std::map<unsigned long, MIB_IPINTERFACE_ROW> returnMapping;
+  PMIB_IPINTERFACE_TABLE interfaces;
+
+  if ((interfaces = getInterfaces(type)) != nullptr) {
+    for (int i = 0; i < (int)interfaces->NumEntries; ++i) {
+      MIB_IPINTERFACE_ROW currentRow = interfaces->Table[i];
+      returnMapping.insert(std::pair<unsigned long, MIB_IPINTERFACE_ROW>(
+          currentRow.InterfaceIndex, currentRow));
+    }
+  }
+
+  return returnMapping;
+}
+
+QueryData genIPRoutes(QueryContext& context) {
+  Row r;
+  QueryData results;
+  unsigned long numEntries = 0;
+  auto interfaces = getInterfaceRowMapping();
+  auto adapters = getAdapterAddressMapping();
+  PMIB_IPFORWARD_TABLE2* ipTable = nullptr;
+  ipTable = (PMIB_IPFORWARD_TABLE2*)(malloc(sizeof(PMIB_IPFORWARD_TABLE2)));
+  GetIpForwardTable2(AF_UNSPEC, ipTable);
+  numEntries = ipTable[0]->NumEntries;
+
+  for (unsigned long i = 0; i < numEntries; ++i) {
+    PIP_ADAPTER_INFO actualAdapter = nullptr;
+    auto currentRow = ipTable[0]->Table[i];
+    std::string ifaceIP;
+    auto addrFamily = currentRow.DestinationPrefix.Prefix.si_family;
+    auto actualInterface = interfaces.at(currentRow.InterfaceIndex);
+    char buf[INET6_ADDRSTRLEN];
```

Comment:
Yeah I can try that. I did try that style for API calls that needed to allocate space via `malloc` but for some reason the memory didn't get allocated properly.

For something like this though that may work!